sdk/assertion_provider_pk.go

```diff
@@ -0,0 +1,206 @@
+package sdk
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+)
+
+const (
+	// KeyAssertionID is the standard identifier for key-based assertions.
+	KeyAssertionID = "assertion-key"
+
+	// KeyAssertionSchema is the schema URI for key-based assertions.
+	// v2 includes assertionSchema claim in JWT binding for security against schema substitution attacks.
+	KeyAssertionSchema = "urn:opentdf:key:assertion:v1"
+
+	// logPrefixLength is the maximum length of signature prefixes in log messages
+	logPrefixLength = 64
+)
+
+type PublicKeyStatement struct {
+	Algorithm string `json:"algorithm"`
+	Key       any    `json:"key"`
+}
+
+type KeyAssertionBinder struct {
+	privateKey AssertionKey
+	publicKey  AssertionKey
+}
+
+type KeyAssertionValidator struct {
+	publicKeys       AssertionVerificationKeys
+	verificationMode AssertionVerificationMode
+}
+
+func NewKeyAssertionBinder(privateKey AssertionKey) *KeyAssertionBinder {
+	return &KeyAssertionBinder{
+		privateKey: privateKey,
+	}
+}
+
+func NewKeyAssertionValidator(publicKeys AssertionVerificationKeys) *KeyAssertionValidator {
+	return &KeyAssertionValidator{
+		publicKeys:       publicKeys,
+		verificationMode: FailFast, // Default to secure mode
+	}
+}
+
+// SetVerificationMode updates the verification mode for this validator.
+// This is typically called by the SDK when registering validators to propagate
+// the global verification mode setting.
+func (p *KeyAssertionValidator) SetVerificationMode(mode AssertionVerificationMode) {
+	p.verificationMode = mode
+}
+
+// Schema returns the schema URI this validator handles.
+// Returns wildcard to match any assertion schema when verification keys are provided.
+func (p *KeyAssertionValidator) Schema() string {
+	return SchemaWildcard
+}
+
+func (p KeyAssertionBinder) Bind(_ context.Context, m Manifest) (Assertion, error) {
+	// Create the public key statement
+	statement := PublicKeyStatement{
+		Algorithm: p.publicKey.Alg.String(),
+		Key:       p.publicKey.Key,
+	}
+
+	jsonBytes, err := json.Marshal(statement)
+	if err != nil {
+		return Assertion{}, fmt.Errorf("failed to marshal public key statement: %w", err)
+	}
+	statementValue := string(jsonBytes)
+
+	// Build the assertion
+	assertion := Assertion{
+		ID:             KeyAssertionID,
+		Type:           BaseAssertion,
+		Scope:          PayloadScope,
+		AppliesToState: Unencrypted,
+		Statement: Statement{
+			Format: StatementFormatJSON,
+			Schema: KeyAssertionSchema,
+			Value:  statementValue,
```

Comment:
The statement value is where the actual json or string value goes for the assertion. 

Therefor we can't stick the `PublicKeyStatement` value in that field. The public key needs to get added to the jws that we build which will have the algorithm in the protected header when using rsa or ecc.  

I also believe there is a `jwk` header that will store the public key too but I would do more research into that. 

Example with system metadata https://github.com/opentdf/platform/blob/main/sdk/assertion.go#L306-L332

Reply:
fixed